enhancements/builds/entitlements.md

```diff
@@ -0,0 +1,477 @@
+alternatives / complementary pieces to a global solution are articulated below.
+
+The preferred implementation path at this time is to build the following set of components inject the global secret contents 
+in a pod:
+ - a [CSI plugin](https://github.com/container-storage-interface/spec) would fetch the secret(s) and write it on the underlying 
+ node/host filesystem on a path we control so that CRIO can mount the content
+ - though for our purposes, `plugin`, `driver`, and `endpoint` seem to be the same thing based on which document you are
+ reading.
+ - k8s provides Kubernetes CSI Sidecar Containers: a set of standard containers that aim to simplify the development and 
+ deployment of CSI Drivers on Kubernetes.
+ - There are CSI volume and persistent volume types in k8s.
+ - per [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#csi) the CSI types do not support direct
+ reference from a Pod and may only be referenced in a Pod via a `PersistentVolumeClaim`
+ - To that end, one of the sidecar containers, the [external provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html),
+ helps with this
+ - It watches Kubernetes `PersistentVolumeClaim` objects, and if the PVC 
+ references a Kubernetes StorageClass, and the name in the provisioner field of the storage class matches the name 
+ returned by the specified CSI endpoint `GetPluginInfo` call, it calls the CSI endpoint `CreateVolume` method to provision
+ the new volume.
+ - there are also lifecycle methods for destroy, resize, etc.  
+ - the CSI plugin would be hosted as a `DaemonSet` on the k8s cluster, to get per node/host granularity
+ - and that `DaemonSet` would be privileged so as to have access to the pod's associated node/host file system.
+ - the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md) details the entire architecture,
+ concepts, and protocol
+ - so Volume provisioning could access the node/host filesystem, in particular the directory (and only the directory)
+ where the secret contents will be copied to
+ - As part of Volume creation/provisioning a key element is this: CSI has two step mounting process. staging and publishing.
+ - They correspond to the endpoint's `NodeStageVolume` and `NodePublishVolume` implementations 
+ - Staging is node local and publish is pod local. So we will probably want to stage the content once and publish it for each pod 
+ (which is nothing but creating a bind mount for original volume)
+ - That means we do not have to necessarily read and copy the secret(s) contents on every pod creation
+ - the [CSI spec volume lifecycle section](https://github.com/container-storage-interface/spec/blob/master/spec.md#volume-lifecycle) 
+ talks about a pre-provisioned volume.
+ - If we can implement this, then we can build an associated controller that watches the global secret, or per namespace secrets that have been 
```

Comment:
The more I read about CSI, the harder it will be IMO to implement something that allows per namespace overrides. I'd prefer namespaced overrides to be a follow-up feature.

Reply:
i would expect the per namespace overrides to occur via the admission controller, not the CSI driver.

Reply:
(But leaving it as a follow up seems totally reasonable)

Reply:
others are orbiting the operator for CSI drivers space ... see https://github.com/openshift/enhancements/pull/242

and +1 on @bparees comment on the per namespace override being on the admissions side vs. CSI driver side

Reply:
also got https://github.com/openshift/enhancements/pull/139 from @gnufied around CSI installation via OLM